realm/realm-library/src/main/cpp/io_realm_internal_OsList.cpp

```diff
@@ -36,28 +41,41 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_OsList_nativeGetFinalizerPtr(JNIE
     return reinterpret_cast<jlong>(&finalize_list);
 }
 
-JNIEXPORT jlongArray JNICALL Java_io_realm_internal_OsList_nativeCreate(JNIEnv* env, jclass, jlong shared_realm_ptr, jlong row_ptr,
-                                                                        jlong column_index)
+JNIEXPORT jlongArray JNICALL Java_io_realm_internal_OsList_nativeCreate(JNIEnv* env, jclass, jlong shared_realm_ptr,
+                                                                        jlong row_ptr, jlong column_index)
 {
     TR_ENTER_PTR(row_ptr)
 
     try {
         auto& row = *reinterpret_cast<realm::Row*>(row_ptr);
 
-        if (!ROW_AND_COL_INDEX_AND_TYPE_VALID(env, &row, column_index, type_LinkList)) {
+        if (!ROW_AND_COL_INDEX_VALID(env, &row, column_index)) {
+            return 0;
+        }
+        auto column_type = row.get_table()->get_column_type(column_index);
+        if (column_type != type_LinkList && column_type != type_Table) {
```

Comment:
Because `CheckedRow.getList()` doesn't check column type.

Other methods in `CheckedRow` calls its native methods and they checks the type of the field, `getList()` just creates `OsList` object and returns it.

We can add another native call to check the type in `CheckedRow.getList()` and it seems a better way because type checking is not necessary when this function is called from generated code.